aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp

```diff
@@ -0,0 +1,98 @@
+#include "init_qnnpack.h"
+
+#include <ATen/ATen.h>
+#include <ATen/core/Type.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/quantized/Quantizer.h>
+#include <ATen/Config.h>
+
+namespace at { namespace native {
+namespace {
+
+class QNNPACKFullyConnected final : public c10::OperatorKernel {
+ public:
+  Tensor operator()(at::Tensor input,
+      at::Tensor weight,
+      at::Tensor bias,
+      double output_scale,
+      int64_t output_zero_point) {
+
+    TORCH_CHECK(input.dim() >= 2, "Input tensor rank should be >= 2");
+
+    Tensor input_contig = input.contiguous();
+
+    // C(output) = A(input_contig) x B(weight), where C, A, B are M x N, M x K, N x K
+    // matrices, respectively.
+    int64_t rows_a = input_contig.size(0);
+    int64_t cols_a = 1;
+    for (size_t i = 1; i < input_contig.dim(); ++i) {
+      cols_a *= input_contig.size(i);
+    }
+
+    int64_t rows_b = weight.size(0);
+
+    TORCH_CHECK(cols_a == weight.size(1),
+        "qnnpack_fc(): input size does not match weight dimension 1 size: got ",
+        cols_a, " but expected ", weight.size(1));
+
+    TORCH_CHECK(!bias.defined() || (bias.ndimension() == 1 && bias.size(0) == rows_b),
+        "qnnpack_fc(): Given weight of size ", weight.sizes(),
+        ", expected bias to be 1-dimensional with ", rows_b, " elements",
+        ", but got bias of size ", bias.sizes(), " instead");
+
+    initQNNPACK();
+    qnnp_operator_t qnnpackOperator_{nullptr};
+
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
+      {rows_a, rows_b},
+      at::device(kCPU).dtype(kQUInt8),
```

Comment:
does qnnpack support only uint8? if so - let's enforce the input dtype below (by using templated data<>), if not - let's use dispatch on dtype

Reply:
yes, updated to use templated data.